source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -413,6 +423,49 @@ DnsCacheImpl::PrimaryHostInfo::~PrimaryHostInfo() {
   parent_.stats_.num_hosts_.dec();
 }
 
+void DnsCacheImpl::addCacheEntry(const std::string& host,
+                                 const Network::Address::InstanceConstSharedPtr& address) {
+  if (!key_value_store_) {
+    return;
+  }
+  // TODO(alyssawilk) cache data should include TTL, or some other indicator.
+  const std::string value = absl::StrCat(address->asString());
+  key_value_store_->addOrUpdate(host, value);
+}
+
+void DnsCacheImpl::removeCacheEntry(const std::string& host) {
+  if (!key_value_store_) {
+    return;
+  }
+  key_value_store_->remove(host);
+}
+
+void DnsCacheImpl::loadCacheEntries(
+    const envoy::extensions::common::dynamic_forward_proxy::v3::DnsCacheConfig& config) {
+  if (!config.has_key_value_config()) {
+    return;
+  }
+  auto& factory =
+      Config::Utility::getAndCheckFactory<KeyValueStoreFactory>(config.key_value_config().config());
+  key_value_store_ = factory.createStore(config.key_value_config(), validation_visitor_,
+                                         main_thread_dispatcher_, file_system_);
+  KeyValueStore::ConstIterateCb load = [this](const std::string& key, const std::string& value) {
+    auto address = Network::Utility::parseInternetAddressAndPortNoThrow(value);
+    if (address == nullptr) {
```

Comment:
Should we log/inc a stat in this case? This means that the cached data is bad right?